From a41656ff8491fe836af15b2cfba280cf1e758365 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 5 Feb 2019 09:34:34 +0100
Subject: [PATCH 2/2] Squash of
 github/lineage-16.0(024702e11)..lineage-16.0-android-9.0.0_r31(9276cdb83) (DO
 NOT SUBMIT)

Ken Chen (1):
 - Free memory when it is no more used

Luca Stefani (1):
 - Merge tag 'android-9.0.0_r31' into lineage-16.0-android-9.0.0_r31

Tim Murray (1):
 - malloc: add M_PURGE mallopt flag

android-build-team Robot (2):
 - Snap for 5087486 from 8a07791fbff0dd014ce5da8d88969d09cd3dcf0b to
   pi-qpr2-release
 - Snap for 5130385 from c466fe95dc641d879683700aea4233ffd42bc762 to
   pi-qpr2-release

Change-Id: I596f324e6673eefb2e770df66e449d0c0182e80f

Change-Id: I280065f8294bfe484c4aee18d7b09aab3395b7f7
---
 libc/dns/resolv/res_cache.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/libc/dns/resolv/res_cache.c b/libc/dns/resolv/res_cache.c
index dda869495..8de9643a3 100644
--- a/libc/dns/resolv/res_cache.c
+++ b/libc/dns/resolv/res_cache.c
@@ -2067,14 +2067,19 @@ _resolv_set_nameservers_for_net(unsigned netid, const char** servers, unsigned n
             // max_samples actually change, in practice the overhead of checking is higher than the
             // cost, and overflows are unlikely
             ++cache_info->revision_id;
-        } else if (cache_info->params.max_samples != old_max_samples) {
-            // If the maximum number of samples changes, the overhead of keeping the most recent
-            // samples around is not considered worth the effort, so they are cleared instead. All
-            // other parameters do not affect shared state: Changing these parameters does not
-            // invalidate the samples, as they only affect aggregation and the conditions under
-            // which servers are considered usable.
-            _res_cache_clear_stats_locked(cache_info);
-            ++cache_info->revision_id;
+        } else {
+            if (cache_info->params.max_samples != old_max_samples) {
+                // If the maximum number of samples changes, the overhead of keeping the most recent
+                // samples around is not considered worth the effort, so they are cleared instead.
+                // All other parameters do not affect shared state: Changing these parameters does
+                // not invalidate the samples, as they only affect aggregation and the conditions
+                // under which servers are considered usable.
+                _res_cache_clear_stats_locked(cache_info);
+                ++cache_info->revision_id;
+            }
+            for (unsigned j = 0; j < numservers; j++) {
+                freeaddrinfo(nsaddrinfo[j]);
+            }
         }
 
         // Always update the search paths, since determining whether they actually changed is
-- 
2.17.1

